### PR TITLE
2015 07 02/tarreader.gzipcmd

### PR DESCRIPTION
**Pull request by hallyn**
Skip the rsync before calling tar, and use archive/tar for creating the tarfile.

This won't yet handle xattrs, which rsync+tar presumably did.

### CODE REVIEW

File: lxd/images.go [hunk 1]
```diff
@@ -167,29 +167,29 @@ func imgPostContInfo(d *Daemon, r *http.Request, req imageFromContainerPostReq,
...
-	args = append(args, "rootfs")
-	output, err := exec.Command("tar", args...).CombinedOutput()
+	if err := c.exportToTar(snap, tarfile); err != nil {
+		tarfile.Close()
+		return info, fmt.Errorf("imgPostContInfo: exportToTar failed: %s\n", err)
+	}
+	tarfile.Close()
+
+	args := []string{tarpath}
+	_, err = exec.Command("gzip", args...).CombinedOutput()
```
Comment (stgraber) [inline lxd/images.go]: Hmm, why move from xz to gzip here?

Comment (stgraber) [inline lxd/images.go]: Also surprised there's no Go-native compression library.

Comment (hallyn) [inline lxd/images.go]: Oh, good point.  I guess bc xz wasn't supported by golang's native compress/* libraries.  May as well go back to using xz indeed.

**Comment by stgraber** [comment]
I also notice you only really take care of a single codepath where we shell out to tar today. Are you planning on converting the others too at some point, or is there no real benefit from doing that?

File: lxd/images.go (hunk 1 above)
Comment (hallyn) [inline lxd/images.go]: Time to publish went from 20s to 3m5s, with file sizes going from 119960845 down to 78430596.  The code using xz is in my 2015-07-03/tarreader.xz branch.

 So let's leave it as gz for now bc 3m to publish is a horrible user experience, and was the motivation for this patch originally.  I'll open an issue for a config key to allow users to choose xz.

Comment (stgraber) [inline lxd/images.go]: ok, fair enough

**Comment by hallyn** [comment]
The goal here wasn't to avoid shelling out to tar for the sake of not shelling out.  It was to avoid needing separate rsync + tar steps which were slow and io-intensive.

We had to do a separate rsync before tar in order to unshift the file uids.  We couldn't unshift in the container rootfs, so we had to either rsync, unshift, then tar, or use archive/tar and unshift in memory.

**Comment by stgraber** [comment]
Ok. As for usual gotchas when dealing with tarballs directly, can you confirm that symlinks are handled properly, in that they are indeed stored as symlinks and the symlink target is never dereferenced?

I'm asking specifically about an absolute symlink in the container which would point to say /etc/shadow, I want to make sure we don't somehow end up reading the host's /etc/shadow.

**Comment by hallyn** [comment]
The symlink target is not dereferenced.  However, I did just find that there is a problem with the uid shifting - if there is any uid which is not valid in the container namespace, it becomes shifted to something which breaks the untar.

I'm actually not sure what we want to do there.  Let's say we have a host-root-owned file accessible in the container.  we dont' want to not unshift it, bc that could lead to a info leak.

So probably we should detect that case and simply ignore the file, leave it out of the published image.

**Comment by stgraber** [comment]
might even be appropriate to just fail for now

**Comment by hallyn** [comment]
True.  The commit I just pushed ignores the file, but if you prefer failing the publish with a clear message i can change it.

**Pull request merged**